### PR TITLE
(new theory) Update TheorySep to new interface

### PR DESCRIPTION
**Pull request by ajreynol**
This updates the theory of separation logic to the new interface, which involves splitting up its check method based on the 4 check callbacks and using the theory state in the standard way.

No behavior changes, unfortunately a lot of code had to change indentation and was updated to new coding guidelines.

### CODE REVIEW

File: src/theory/sep/theory_sep.cpp [hunk 1]
```diff
@@ -1486,26 +1507,34 @@ void TheorySep::setInactiveAssertionRec( Node fact, std::map< Node, std::vector<
...
+  for (unsigned i = 0, nchild = satom.getNumChildren(); i < nchild; i++)
+  {
+    Node lblc = getLabel(satom, i, lbl);
     Assert(!lblc.isNull());
     std::map< Node, Node > visited;
-    Node lc = applyLabel( s_atom[i], lblc, visited );
+    Node lc = applyLabel(satom[i], lblc, visited);
     Assert(!lc.isNull());
-    if( i==1 && s_atom.getKind()==kind::SEP_WAND ){
+    if (i == 1 && satom.getKind() == kind::SEP_WAND)
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:1536]:
```suggestion
    if (i == 1 && satom.getKind() == SEP_WAND)
```

File: src/theory/sep/theory_sep.cpp [hunk 2]
```diff
@@ -1486,26 +1507,34 @@ void TheorySep::setInactiveAssertionRec( Node fact, std::map< Node, std::vector<
...
 
-void TheorySep::getLabelChildren( Node s_atom, Node lbl, std::vector< Node >& children, std::vector< Node >& labels ) {
-  for( unsigned i=0; i<s_atom.getNumChildren(); i++ ){
-    Node lblc = getLabel( s_atom, i, lbl );
+void TheorySep::getLabelChildren(Node satom,
+                                 Node lbl,
+                                 std::vector<Node>& children,
+                                 std::vector<Node>& labels)
+{
+  for (unsigned i = 0, nchild = satom.getNumChildren(); i < nchild; i++)
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:1529]:
```suggestion
  for (size_t i = 0, nchild = satom.getNumChildren(); i < nchild; i++)
```

File: src/theory/sep/theory_sep.cpp [hunk 3]
```diff
@@ -1486,26 +1507,34 @@ void TheorySep::setInactiveAssertionRec( Node fact, std::map< Node, std::vector<
...
   bool polarity = fact.getKind() != kind::NOT;
   TNode atom = polarity ? fact : fact[0];
-  TNode s_atom = atom[0];
-  TNode s_lbl = atom[1];
-  if( s_atom.getKind()==kind::SEP_WAND || s_atom.getKind()==kind::SEP_STAR ){
-    for( unsigned j=0; j<s_atom.getNumChildren(); j++ ){
-      Node lblc = getLabel( s_atom, j, s_lbl );
+  TNode satom = atom[0];
+  TNode slbl = atom[1];
+  if (satom.getKind() == kind::SEP_WAND || satom.getKind() == kind::SEP_STAR)
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:1512]:
```suggestion
  if (satom.getKind() == SEP_WAND || satom.getKind() == SEP_STAR)
```

File: src/theory/sep/theory_sep.cpp [hunk 4]
```diff
@@ -312,219 +278,250 @@ void TheorySep::presolve() {
...
+bool TheorySep::isSpatialKind(Kind k) const
+{
+  return k == SEP_STAR || k == SEP_WAND || k == SEP_PTO || k == SEP_EMP;
+}
+
+void TheorySep::postCheck(Effort level)
+{
+  if (level == EFFORT_LAST_CALL && !d_state.isInConflict()
+      && !d_valuation.needCheck())
+  {
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:524]: Can we refactor the function to use less indentation by doing something like:

```
if (level != EFFORT_LAST_CALL || d_state.isInConflict() || d_valuation.needCheck())
{
  return;
}
```

File: src/theory/sep/theory_sep.cpp [hunk 5]
```diff
@@ -312,219 +278,250 @@ void TheorySep::presolve() {
...
-      //add to spatial assertions
-      if( !d_conflict && is_spatial ){
-        d_spatial_assertions.push_back( fact );
+      // send out definitional lemmas for introduced sets
+      for (const Node& clem : c_lems)
+      {
+        Trace("sep-lemma") << "Sep::Lemma : definition : " << clem << std::endl;
+        d_out->lemma(clem);
       }
+      // children.insert( children.end(), c_lems.begin(), c_lems.end() );
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:432]:
```suggestion
```

File: src/theory/sep/theory_sep.cpp [hunk 6]
```diff
@@ -312,219 +278,250 @@ void TheorySep::presolve() {
...
+      if (slbl != ss)
+      {
+        conc = slbl.eqNode(ss);
+      }
+      // note semantics of sep.nil is enforced globally
+    }
+    else if (satom.getKind() == SEP_EMP)
+    {
+      // conc = slbl.eqNode(
+      // nm->mkConst(EmptySet(slbl.getType())) );
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:447]:
```suggestion
```

File: src/theory/sep/theory_sep.cpp [hunk 7]
```diff
@@ -312,219 +278,250 @@ void TheorySep::presolve() {
...
+        // Node econc = nm->mkNode( AND, slbl.eqNode( emp_s ).negate(),
+        lem = nm->mkNode(OR, fact.negate(), econc);
+      }
+      Trace("sep-lemma") << "Sep::Lemma : emp : " << lem << std::endl;
+      d_out->lemma(lem);
+    }
+    else
+    {
+      // labeled emp should be rewritten
+      Assert(false);
```
Comment (4tXJ7f) [inline src/theory/sep/theory_sep.cpp:471]:
```suggestion
      Unreachable();
```

File: src/theory/sep/theory_sep.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/sep/theory_sep.cpp:524]: Sure, was trying to avoid some indentation changes to keep the LOC changes small but I can change this.

**Pull request merged**